x/sisu/handler_contract_change_liquidity_address.go

```diff
@@ -0,0 +1,138 @@
+	libchain "github.com/sisu-network/lib/chain"
+	"github.com/sisu-network/lib/log"
+	"github.com/sisu-network/sisu/common"
+	"github.com/sisu-network/sisu/x/sisu/keeper"
+	"github.com/sisu-network/sisu/x/sisu/tssclients"
+	"github.com/sisu-network/sisu/x/sisu/types"
+)
+
+type HandlerContractChangeLiquidityAddress struct {
+	pmm      PostedMessageManager
+	publicDb keeper.Storage
+	mc       ManagerContainer
+}
+
+func NewHandlerContractChangeLiquidityAddress(mc ManagerContainer) *HandlerContractChangeLiquidityAddress {
+	return &HandlerContractChangeLiquidityAddress{
+		publicDb: mc.PublicDb(),
+		pmm:      mc.PostedMessageManager(),
+		mc:       mc,
+	}
+}
+
+func (h *HandlerContractChangeLiquidityAddress) DeliverMsg(ctx sdk.Context, msg *types.ChangeLiquidPoolAddressMsg) (*sdk.Result, error) {
+	if process, hash := h.pmm.ShouldProcessMsg(ctx, msg); process {
+		newHandlerContractChangeLiquidityAddress(h.mc).doChangeLiquidityAddress(ctx, msg.Data.Chain, msg.Data.Hash, msg.Data.NewLiquidAddress)
+		h.publicDb.ProcessTxRecord(hash)
+	} else {
+		log.Verbose("HandlerContractChangeOwnership: didn't not reach consensus or transaction has been processed")
+	}
+
+	return &sdk.Result{}, nil
+}
+
+type handlerContractChangeLiquidityAddress struct {
+	publicDb         keeper.Storage
+	txOutputProducer TxOutputProducer
+	globalData       common.GlobalData
+	partyManager     PartyManager
+	dheartClient     tssclients.DheartClient
+}
+
+func newHandlerContractChangeLiquidityAddress(mc ManagerContainer) *handlerContractChangeLiquidityAddress {
+	return &handlerContractChangeLiquidityAddress{
+		publicDb:         mc.PublicDb(),
+		txOutputProducer: mc.TxOutProducer(),
+		globalData:       mc.GlobalData(),
+		partyManager:     mc.PartyManager(),
+		dheartClient:     mc.DheartClient(),
+	}
+}
+
+func (h *handlerContractChangeLiquidityAddress) doChangeLiquidityAddress(ctx sdk.Context, chain, hash, newLpAddress string) ([]byte, error) {
+	// Only do pause/pause if we finished catching up.
+	if h.globalData.IsCatchingUp() {
+		log.Info("We are catching up with the network, exiting doPauseOrResume")
+		return nil, nil
+	}
+
+	found := false
+	for _, contract := range SupportedContracts {
+		if strings.EqualFold(strings.ToLower(contract.AbiHash), strings.ToLower(hash)) {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		err := fmt.Errorf("doPauseOrResume: contract with hash %s is not supported", hash)
+		log.Error(err)
+		return nil, err
+	}
+
+	txOutMsg, err := h.txOutputProducer.ContractChangeLiquidPoolAddress(ctx, chain, hash, newLpAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// Save this to KVStore
+	h.publicDb.SaveTxOut(txOutMsg.Data)
+
+	// Sends to dheart for signing.
+	h.signTx(ctx, txOutMsg.Data)
+
+	return nil, nil
+}
+
+// TODO: duplicate code with pause/resume contract handler, fix it
+// signTx sends a TxOut to dheart for TSS signing.
+func (h *handlerContractChangeLiquidityAddress) signTx(ctx sdk.Context, tx *types.TxOut) {
```

Comment:
This code sends a TxOut to dheart to sign. Seems like this duplicates in several places in the codebase. Worth doing a refactoring in this PR.